// redux-draft/lib/constants/ActionTypes.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
var INIT = exports.INIT = 'RD/INIT';
var SET_REF = exports.SET_REF = 'RD/SET_REF';
var UPDATE_EDITOR_STATE = exports.UPDATE_EDITOR_STATE = 'RD/UPDATE_EDITOR_STATE';
var UPDATE_EDITOR_STATE_FROM_RAW = exports.UPDATE_EDITOR_STATE_FROM_RAW = 'RD/UPDATE_EDITOR_STATE_FROM_RAW';
var UPDATE_READ_ONLY = exports.UPDATE_READ_ONLY = 'RD/UPDATE_READ_ONLY';
var TOGGLE_BLOCK = exports.TOGGLE_BLOCK = 'RD/TOGGLE_BLOCK';
var TOGGLE_STYLE = exports.TOGGLE_STYLE = 'RD/TOGGLE_STYLE';
var APPLY_BLOCK = exports.APPLY_BLOCK = 'RD/APPLY_BLOCK';
var REMOVE_ENTITY = exports.REMOVE_ENTITY = 'RD/REMOVE_ENTITY';
var APPLY_ENTITY = exports.APPLY_ENTITY = 'RD/APPLY_ENTITY';
var INSERT_ENTITY = exports.INSERT_ENTITY = 'RD/INSERT_ENTITY';
var INSERT_ATOMIC_BLOCK = exports.INSERT_ATOMIC_BLOCK = 'RD/INSERT_ATOMIC_BLOCK';
var REMOVE_BLOCK = exports.REMOVE_BLOCK = 'RD/REMOVE_BLOCK';
var UPDATE_ENTITY_DATA = exports.UPDATE_ENTITY_DATA = 'RD/UPDATE_ENTITY_DATA';
