// redux-draft/lib/actions/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.updateEntityData = exports.removeBlock = exports.insertAtomicBlock = exports.insertEntity = exports.applyEntity = exports.removeEntity = exports.applyBlock = exports.toggleStyle = exports.toggleBlock = exports.updateReadOnly = exports.updateEditorStateFromRaw = exports.updateEditorState = exports.setRef = exports.init = undefined;

var _ActionTypes = require('../constants/ActionTypes');

var init = exports.init = function init(editorName, config) {
  var markInitialized = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : true;

  return {
    type: _ActionTypes.INIT,
    editorName: editorName,
    config: config,
    markInitialized: markInitialized
  };
};

var setRef = exports.setRef = function setRef(editorName, editorInstance) {
  return {
    type: _ActionTypes.SET_REF,
    editorName: editorName,
    editorInstance: editorInstance
  };
};

var updateEditorState = exports.updateEditorState = function updateEditorState(editorName, editorState) {
  return {
    type: _ActionTypes.UPDATE_EDITOR_STATE,
    editorName: editorName,
    editorState: editorState
  };
};

var updateEditorStateFromRaw = exports.updateEditorStateFromRaw = function updateEditorStateFromRaw(editorName, rawContent) {
  return {
    type: _ActionTypes.UPDATE_EDITOR_STATE_FROM_RAW,
    editorName: editorName,
    rawContent: rawContent
  };
};

var updateReadOnly = exports.updateReadOnly = function updateReadOnly(editorName, readOnly) {
  return {
    type: _ActionTypes.UPDATE_READ_ONLY,
    editorName: editorName,
    readOnly: readOnly
  };
};

var toggleBlock = exports.toggleBlock = function toggleBlock(editorName, blockName) {
  return {
    type: _ActionTypes.TOGGLE_BLOCK,
    editorName: editorName,
    blockName: blockName
  };
};

var toggleStyle = exports.toggleStyle = function toggleStyle(editorName, styleName) {
  return {
    type: _ActionTypes.TOGGLE_STYLE,
    editorName: editorName,
    styleName: styleName
  };
};

var applyBlock = exports.applyBlock = function applyBlock(editorName, blockName) {
  return {
    type: _ActionTypes.APPLY_BLOCK,
    editorName: editorName,
    blockName: blockName
  };
};

var removeEntity = exports.removeEntity = function removeEntity(editorName) {
  return {
    type: _ActionTypes.REMOVE_ENTITY,
    editorName: editorName
  };
};

var applyEntity = exports.applyEntity = function applyEntity(editorName, entityName, mutability, data) {
  return {
    type: _ActionTypes.APPLY_ENTITY,
    editorName: editorName,
    entityName: entityName,
    mutability: mutability,
    data: data
  };
};

var insertEntity = exports.insertEntity = function insertEntity(editorName, entityName, mutability, data) {
  var text = arguments.length > 4 && arguments[4] !== undefined ? arguments[4] : ' ';

  return {
    type: _ActionTypes.INSERT_ENTITY,
    editorName: editorName,
    entityName: entityName,
    mutability: mutability,
    data: data,
    text: text
  };
};

var insertAtomicBlock = exports.insertAtomicBlock = function insertAtomicBlock(editorName, entityName, mutability, data) {
  var text = arguments.length > 4 && arguments[4] !== undefined ? arguments[4] : ' ';

  return {
    type: _ActionTypes.INSERT_ATOMIC_BLOCK,
    editorName: editorName,
    entityName: entityName,
    mutability: mutability,
    data: data,
    text: text
  };
};

var removeBlock = exports.removeBlock = function removeBlock(editorName, blockKey) {
  return {
    type: _ActionTypes.REMOVE_BLOCK,
    editorName: editorName,
    blockKey: blockKey
  };
};

var updateEntityData = exports.updateEntityData = function updateEntityData(editorName, entityKey, data) {
  return {
    type: _ActionTypes.UPDATE_ENTITY_DATA,
    editorName: editorName,
    entityKey: entityKey,
    data: data
  };
};
