// redux-draft/lib/components/reduxDraft.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _redux = require('redux');

var _reactRedux = require('react-redux');

var _index = require('../actions/index');

var actionCreators = _interopRequireWildcard(_index);

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) newObj[key] = obj[key]; } } newObj.default = obj; return newObj; } }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

exports.default = function (config) {
  return function (WrappedComponent) {
    var ReduxDraft = function (_Component) {
      _inherits(ReduxDraft, _Component);

      function ReduxDraft(props) {
        _classCallCheck(this, ReduxDraft);

        var _this = _possibleConstructorReturn(this, (ReduxDraft.__proto__ || Object.getPrototypeOf(ReduxDraft)).call(this, props));

        var _initialized = props._initialized,
            init = props.init;


        if (!_initialized) {
          init(config);
        }
        return _this;
      }

      _createClass(ReduxDraft, [{
        key: 'getChildContext',
        value: function getChildContext() {
          var _props = this.props,
              updateEditorState = _props.updateEditorState,
              updateReadOnly = _props.updateReadOnly,
              updateEntityData = _props.updateEntityData,
              removeBlock = _props.removeBlock;


          return {
            _reduxDraft: {
              updateEditorState: updateEditorState,
              updateReadOnly: updateReadOnly,
              updateEntityData: updateEntityData,
              removeBlock: removeBlock
            }
          };
        }
      }, {
        key: 'render',
        value: function render() {
          var _props2 = this.props,
              _instance = _props2._instance,
              name = _props2.name,
              config = _props2.config,
              editorState = _props2.editorState,
              customStyleMap = _props2.customStyleMap,
              rest = _objectWithoutProperties(_props2, ['_instance', 'name', 'config', 'editorState', 'customStyleMap']);

          return name !== undefined ? _react2.default.createElement(WrappedComponent, _extends({
            focus: (_instance || {}).focus,
            name: name,
            config: config,
            editorState: editorState,
            customStyleMap: customStyleMap
          }, rest)) : null;
        }
      }]);

      return ReduxDraft;
    }(_react.Component);

    ReduxDraft.childContextTypes = {
      _reduxDraft: _propTypes2.default.object
    };

    var mapStateToProps = function mapStateToProps(state) {
      var currentDraft = state.draft[config.name] || {};

      return currentDraft;
    };
    var mapDispatchToProps = function mapDispatchToProps(dispatch) {
      var bindDraft = function bindDraft(actionCreator) {
        return actionCreator.bind(null, config.name);
      };
      var bondDraftActionCreators = {};

      Object.keys(actionCreators).forEach(function (ac) {
        bondDraftActionCreators[ac] = bindDraft(actionCreators[ac]);
      });

      return _extends({}, (0, _redux.bindActionCreators)(bondDraftActionCreators, dispatch), {
        dispatch: dispatch
      });
    };

    return (0, _reactRedux.connect)(mapStateToProps, mapDispatchToProps)(ReduxDraft);
  };
};
